[MRG] ENH: dataset-fetching with use figshare, partial download and checksum

nelson-liu opened the pull request:
#### Reference Issue

addresses #7425, #8089 
#### What does this implement/fix? Explain your changes.

- I added all of the datasets (`fetch_...`) to figshare and changed their URLs accordingly.
- Also added a custom download function that enables users to resume interrupted downloads
- Lastly, this PR implements md5 checking of both the raw downloaded files and the produced datasets

#### Any other comments?

I'm pretty sure CI doesn't test datasets that aren't already downloaded to my machine, is there a way to test this PR besides going into all of the tests and setting `download_if_missing=True`?

@amueller I emailed the figshare details to your gmail (t3kcit)

nelson-liu committed:
add 20newsgroups dataset to figshare

nelson-liu committed:
made link less verbose

nelson-liu committed:
add olivetti to figshare

nelson-liu committed:
add lfw to figshare

nelson-liu committed:
add california housing dataset to figshare

nelson-liu committed:
add covtype dataset to figshare

nelson-liu committed:
add kddcup99 dataset to figshare

nelson-liu committed:
add species distribution dataset to figshare

nelson-liu committed:
add rcv1 dataset

amueller commented:
sorry for not following up on this. Will review shortly.

nelson-liu commented:
@amueller no problem, this sort of thing is pretty tedious to check.

jnothman commented on sklearn/datasets/covtype.py:
```diff
@@ -33,8 +33,7 @@
 
-URL = ('http://archive.ics.uci.edu/ml/'
-       'machine-learning-databases/covtype/covtype.data.gz')
+URL = ('https://ndownloader.figshare.com/files/5976039')
```
no need to keep parentheses. They were there for line-wrapped strings only.

jnothman commented on sklearn/datasets/lfw.py:
```diff
@@ -44,14 +44,15 @@
 logger = logging.getLogger(__name__)
 
 
-BASE_URL = "http://vis-www.cs.umass.edu/lfw/"
```
Hmm. Can we assume no backwards compatibility issues with these names? I'm fine with this.

nelson-liu replied:
hmm, I believe so. I specifically made an attempt to maintain backwards compatibility by changing the associated variable names

jnothman commented on sklearn/datasets/kddcup99.py:
```diff
@@ -29,11 +29,9 @@
 
-URL10 = ('http://archive.ics.uci.edu/ml/'
-         'machine-learning-databases/kddcup99-mld/kddcup.data_10_percent.gz')
+URL10 = ('https://ndownloader.figshare.com/files/5976042')
```
parens

jnothman commented on sklearn/datasets/kddcup99.py:
```diff
@@ -29,11 +29,9 @@
 
-URL = ('http://archive.ics.uci.edu/ml/'
-       'machine-learning-databases/kddcup99-mld/kddcup.data.gz')
+URL = ('https://ndownloader.figshare.com/files/5976045')
```
parens

jnothman commented on sklearn/datasets/rcv1.py:
```diff
@@ -28,10 +28,14 @@
+    'https://ndownloader.figshare.com/files/5976060',
+    'https://ndownloader.figshare.com/files/5976057'
+]
+URL_topics = ('https://ndownloader.figshare.com/files/5976048')
```
parens

jnothman commented on sklearn/datasets/twenty_newsgroups.py:
```diff
@@ -64,8 +64,7 @@
 
-URL = ("http://people.csail.mit.edu/jrennie/"
-       "20Newsgroups/20news-bydate.tar.gz")
+URL = ("https://ndownloader.figshare.com/files/5975967")
```
parens

jnothman reviewed (changes requested):
From a code perspective this otherwise LGTM

nelson-liu committed:
remove extraneous parens from url strings

jnothman commented:
```
echo '
from sklearn.datasets import *
for name in locals().keys():
     if name.startswith("fetch_") and name != "fetch_mldata":
         print(name)
         locals()[name]()
' > /tmp/foo.py &&
mkdir /tmp/master_data /tmp/pr_data &&
git checkout master &&
SCIKIT_LEARN_DATA=/tmp/master_data python /tmp/foo.py &&
git checkout upstream/pr/7429 &&
SCIKIT_LEARN_DATA=/tmp/pr_data python /tmp/foo.py
```

I've then compared the MD5 sums produced and found them identical except for generated images. 

However, I've realised that those MD5 sum checks should be part of the fetch.

Please add this:
```
05297c95ca4efd1017820d17e358a11e  /tmp/master_data/kddcup99_10/samples
0888a9e9b09a04af8c5d542973ed1950  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_pairs/4dacd9a65737d5fda3ff6c02e3dd167b/output.pkl
0a761020de3d898d00cdb69e37ada346  /tmp/master_data/lfw_home/lfw_funneled/pairs_05.txt
0e896d78ea4c95b72b2914089e9503a6  /tmp/master_data/RCV1/sample_topics.pkl
0f97d450ebcc31d9489bc40ea7db1de1  /tmp/master_data/RCV1/sample_id.pkl
10ff5148d371ea69a620692665553612  /tmp/master_data/lfw_home/lfw_funneled/pairs_06.txt
25455304fc74c07fd031297c8fffd4e8  /tmp/master_data/kddcup99_10/targets
2a5f838e615f5423947945a28a3ef5bd  /tmp/master_data/olivetti.pkz
321690c7b33b99fdfbcf20caabc9fd32  /tmp/master_data/lfw_home/lfw_funneled/pairs_01.txt
3a17d79fe25a12233872ae721603020c  /tmp/master_data/lfw_home/lfw_funneled/pairs_08.txt
3c19f7377094e966cc5ee26f1ed40534  /tmp/master_data/lfw_home/lfw_funneled/pairs_02.txt
41e98d056b196a9ffe00b09c2565e162  /tmp/master_data/covertype/samples_01.npy.z
4f27cbf15b2da4a85c1907eb4181ad21  /tmp/master_data/lfw_home/pairsDevTrain.txt
5132f7440eb68cf58910c8a45a2ac10b  /tmp/master_data/lfw_home/pairsDevTest.txt
564261c7565ce0b57ed44f1f77ff0ea0  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_people/040ffd485691ec83d5af84299b39ea70/output.pkl_01.npy.z
5739836ebdbe6adaf6a4e499ee6fd6d1  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_pairs/func_code.py
5a05ba128212f5b4169f7ee2f460e81a  /tmp/master_data/RCV1/samples.pkl_01.npy.z
64ab790f52560900b85e4869eaf61c15  /tmp/master_data/lfw_home/lfw_funneled/pairs_09.txt
64bbe259a6e08b050332b74809e8f0ee  /tmp/master_data/species_coverage.pkz
78fbf6d80131f3c66e5cc7ac0ddc03c2  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_people/func_code.py
88012b0d731694b416dec9b85ca3c800  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_people/040ffd485691ec83d5af84299b39ea70/metadata.json
98b8efef98c2c3602517ef6fb2990539  /tmp/master_data/RCV1/topics_names.pkl
9d15045fb2719e6b19dd158972982b13  /tmp/master_data/lfw_home/lfw_funneled/pairs_04.txt
9f1ba174e4e1c508ff7cdf10ac338a7d  /tmp/master_data/lfw_home/lfw_funneled/pairs.txt
9f1ba174e4e1c508ff7cdf10ac338a7d  /tmp/master_data/lfw_home/pairs.txt
a0d5d66ebb457948fd502f71d7ffdb6e  /tmp/master_data/lfw_home/lfw_funneled/pairs_03.txt
a5420a1cae0b2270d33fec1075341d38  /tmp/master_data/covertype/samples
a9d6eb2c2b080e238b89cc996eb328af  /tmp/master_data/RCV1/samples.pkl
b23cf9a382bc3a9d8352ae3fdc98b878  /tmp/master_data/covertype/targets
b6911731dd9f956d1f4df612fd127538  /tmp/master_data/RCV1/samples.pkl_02.npy.z
be6dacd276a598cd17755331054ca7b1  /tmp/master_data/lfw_home/lfw_funneled/pairs_10.txt
d7e080432f2e9ba4f1d3d6171f6113c8  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_pairs/4dacd9a65737d5fda3ff6c02e3dd167b/metadata.json
d8aff5fdc363ff82950145dabd2343bd  /tmp/master_data/20newsgroup_vectorized.pkl
e08bb31595a445c56a991c0da7d85e83  /tmp/master_data/cal_housing.pkz
eeab2af266f3bfb0767b7ea2d7fafe3c  /tmp/master_data/lfw_home/lfw_funneled/pairs_07.txt
f445b493da79987ef5cb2b26d09e2f88  /tmp/master_data/20news-bydate.pkz
f6c3e4ff0acc44604bc7efb457559775  /tmp/master_data/lfw_home/joblib/sklearn/datasets/lfw/_fetch_lfw_people/040ffd485691ec83d5af84299b39ea70/output.pkl
```

I'm also concerned about what we lose by throwing away the original URLs. Should we use a mapping rather than replacing them?

nelson-liu commented:
@jnothman good idea, but I'm not too sure where you want me to add the MD5 checks to. Could you elaborate a bit more

> I'm also concerned about what we lose by throwing away the original URLs. Should we use a mapping rather than replacing them?

Hmm...perhaps I'm being slow, but how would such a mapping work?

jnothman commented:
I would create a helper function that fetches a file from a particular URL and stores it at a path relative to data home, with an MD5 check. So it would be passed the URL       and the expected MD5 among other args. Not sure if that's sufficient.

jnothman commented:
Not sure if mapping is necessary

nelson-liu commented:
@jnothman what i ended up doing instead was running md5 on the downloaded archives rather than the raw datafiles. I basically just wrote a general `fetch_dataset` function that given a URL, path, and md5, downloads the URL to a temp location, checks the archive / raw data with the given MD5, then moves it to the proper location. Do you think this would be suitable, or is it greatly preferred to run md5 on the extracted files?

jnothman commented:
That's okay, except that it doesn't catch the case where files later get
corrupted.

On 21 December 2016 at 19:07, Nelson Liu <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> what i ended up doing instead was
> running md5 on the downloaded archives rather than the raw datafiles. I
> basically just wrote a general fetch_dataset function that given a URL,
> path, and md5, downloads the URL to a temp location, checks the archive /
> raw data with the given MD5, then moves it to the proper location. Do you
> think this would be suitable, or is it greatly preferred to run md5 on the
> extracted files?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7429#issuecomment-268461133>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68U4T8loSOcrupef-rLFvo-4s0Iaks5rKN49gaJpZM4J9Y_O>
> .
>

nelson-liu commented:
@jnothman i initially had written a function to just check the archives, but I decided to check the hashes of the generated files as well. However, it seems like i get different md5 hashes than you do. For example, `cal_housing.pkz` has an md5 hash of `39c2dc70c4aad72e44b741c37163e6cc` on my machine as calculated by `md5sum`...

nelson-liu committed:
check md5 of datasets and add resume functionality to downloads

nelson-liu committed:
remove extraneous print statements

nelson-liu commented:
Got a chance to revamp some work I was doing on this PR, here's a summary of what it does:

- I added all of the datasets (`fetch_...`) to figshare and changed their URLs accordingly.
- Also added a custom download function that enables users to resume interrupted downloads
- Lastly, this PR implements md5 checking of both the raw downloaded files and the produced datasets

I'm away from stable internet at the moment, so I'll wait to do extensive testing of this PR until the day after tomorrow. Due to my spotty connection, I've been able to naturally test this a few times 😉 i ran each fetching function after modifying it to make sure it still works, but I want to write a script to automatically go through and verify that all the datasets are in proper working order.

nelson-liu commented:
i'll also provide a script / some data here that compares the md5s of the files after this PR and the current master, to ensure that they're the same.

nelson-liu committed:
fix flake8 violations

nelson-liu committed:
add docstrings to new dataset fetching functions

nelson-liu committed:
consolidate imports in base and use md5 check function in dl

nelson-liu committed:
remove accidentally removed import

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
     return join(*new_args)
+
+
+class partialURLOpener(urllib.FancyURLopener):
```
start with uppercase P

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+    """
+
+    md5hash = hashlib.md5()
+    md5hash.update(open(path, 'rb').read())
```
You might want to read and update in chunks if we're dealing with any big datasets.

nelson-liu replied:
yeah definitely, good suggestion.

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+
+def md5(path):
+    """
+    Calculate the md5 hash of the file at path.
```
PEP257: this summary belongs on the previous line

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+
+
+class partialURLOpener(urllib.FancyURLopener):
+    """
```
PEP257: describe the class here

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+
+
+def validate_file_md5(expected_checksum, path):
+    """
```
PEP257: one-line description goes here

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+
+    existing_size = 0
+    resume_url_downloader = partialURLOpener()
+    path_temp = path + ".tmp"
```
Perhaps a more descriptive extension, like ".part" is appropriate

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+        print("Resuming download from previous temp file, "
+              "already have {} bytes".format(existing_size))
+        # Download only the remainder of the file
+        resume_url_downloader.addheader("Range", "bytes="
```
IIRC, servers are not required to support this, which may mean you're sent the file from scratch.

I wonder whether implementing resuming is a bit much.

nelson-liu replied:
>IIRC, servers are not required to support this, which may mean you're sent the file from scratch.
This is true, but it's part of the original HTTP/1.1 spec, which means that (virtually all?) most webservers support it. Beyond that, i do agree that its a good idea to implement a check to see if the header is supported and act accordingly

> I wonder whether implementing resuming is a bit much.
It could be very well be, I figured why not use the temp file if we're writing one anyway.

jnothman commented on sklearn/datasets/kddcup99.py:
```diff
@@ -347,6 +345,21 @@ def _fetch_brute_kddcup99(subset=None, data_home=None,
         joblib.dump(X, samples_path, compress=0)
         joblib.dump(y, targets_path, compress=0)
+
+        # check md5 of dumped samples and targets
```
Hmm... I'm not sure if there's any guarantee about the consistency of a joblib dumped object's binary representation, from run to run or from version to version.

nelson-liu replied:
i see, would it be better to just not check md5 of joblib dumped files?

jnothman replied:
What do we lose by not checking?

nelson-liu replied:
well we'd lose the ability to detect the case where files later get corrupted? or did i misunderstand your original comment?

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
+        pass
+
+
+def md5(path):
```
should probably be private. perhaps the functions below, also.

nelson-liu committed:
attempt to fix docstring conventions / handle case where range header not supported

nelson-liu committed:
change functions to used renamed, privatized utilities

nelson-liu commented:
thanks for the thorough review @jnothman , I tried to address the issues you brought up. Let me know if you see something lacking.

nelson-liu committed:
fix flake8 indentation error

jnothman commented:
Yes, I suppose that's all. No big deal, I guess.

On 27 December 2016 at 10:14, Nelson Liu <notifications@github.com> wrote:

> *@nelson-liu* commented on this pull request.
> ------------------------------
>
> In sklearn/datasets/kddcup99.py
> <https://github.com/scikit-learn/scikit-learn/pull/7429>:
>
> > @@ -347,6 +345,21 @@ def _fetch_brute_kddcup99(subset=None, data_home=None,
>
>          joblib.dump(X, samples_path, compress=0)
>          joblib.dump(y, targets_path, compress=0)
> +
> +        # check md5 of dumped samples and targets
>
> well we'd lose the ability to detect the case where files later get
> corrupted? or did i misunderstand your original comment?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7429>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6x7PV1btJA05u3Q9dBCsXFmL3mJHks5rMEpYgaJpZM4J9Y_O>
> .
>

nelson-liu committed:
remove checks for joblib dumped files

nelson-liu commented on sklearn/datasets/twenty_newsgroups.py:
```diff
@@ -103,6 +90,10 @@ def download_20newsgroups(target_dir, cache_path):
     with open(cache_path, 'wb') as f:
         f.write(compressed_content)
 
+    # check md5 of written file
```
@jnothman do you know if pickle makes any guarantees about file consistency?

nelson-liu committed:
fix error in lfw

ageron commented:
@nelson-liu This looks great, thanks! Considering mldata.org is down (#8588), can we get this merged quickly? It looks like some conflicts need to be resolved first.

jnothman commented:
I agree. We should aim to merge this soon and certainly ensure it's available for the next release. Are you able to wrap it up @nelson-liu? Fix merge conflicts and I'll give it a final look.

nelson-liu commented:
When we're you planning for the next release? In a bit of a crunch right now, but I should be able to get to this next weekend (unfortunately not the coming one)

jnothman commented:
Current vague release plan is "before June". I hope we can actually make it
happen. So this is urgent only from the perspective of users seeking
solutions in master to current broken functionality.

We can get someone else to put in any finishing touches if you'd rather.

On 5 April 2017 at 09:11, Nelson Liu <notifications@github.com> wrote:

> When we're you planning for the next release? In a bit of a crunch right
> now, but I should be able to get to this next weekend (unfortunately not
> the coming one)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7429#issuecomment-291667597>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63N-9sahTIQMWM7p7jTNkdTWXYBMks5rss4ngaJpZM4J9Y_O>
> .
>

jnothman commented:
Bump?

nelson-liu commented:
Oops, this fell of my radar; I'll do it today. Thanks for the reminder @jnothman

nelson-liu committed:
Merge branch 'master' into use_figshare_in_datasets

nelson-liu committed:
Add missing Bunch import in california housing

nelson-liu commented:
@jnothman ok, I think it should be good now (pending CI)...I manually reran some of the `fetch_*` functions, and was thinking that perhaps it'd be good to have some more logging messages on them (since i spent a lot of time waiting, and wasn't quite sure what it was doing). That's probably out of scope of this PR though...

lesteve commented:
CircleCI seems genuine:

```
Unexpected failing examples:
../examples/applications/plot_model_complexity_influence.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/ubuntu/scikit-learn/doc/sphinxext/sphinx_gallery/gen_rst.py", line 450, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 2, in <module>
  File "<string>", line 13, in generate_data
  File "/home/ubuntu/scikit-learn/sklearn/datasets/twenty_newsgroups.py", line 328, in fetch_20newsgroups_vectorized
    remove=remove)
  File "/home/ubuntu/scikit-learn/sklearn/datasets/twenty_newsgroups.py", line 215, in fetch_20newsgroups
    cache_path=cache_path)
  File "/home/ubuntu/scikit-learn/sklearn/datasets/twenty_newsgroups.py", line 94, in download_20newsgroups
    _validate_file_md5(expected_checksum, cache_path)
  File "/home/ubuntu/scikit-learn/sklearn/datasets/base.py", line 881, in _validate_file_md5
    "corrupted.".format(path))
ValueError: /home/ubuntu/scikit_learn_data/20news-bydate.pkz has an MD5 hash differing from expected, file may be corrupted.
```

nelson-liu commented:
@lesteve thanks! I think it'd be best to remove the validation on the actual processed file, since I also realized that the hash of the generated file would change depending on python version (2 vs 3)...

nelson-liu committed:
Remove hash validation of 20news output pkl

nelson-liu committed:
Remove unused import

nelson-liu commented:
appveyor failure looks spurious...does someone want to rerun that test or should i just push a spurious commit?

jnothman commented:
I'm sure we'll find something to nitpick about to make your next commit not spurious.

jnothman commented on sklearn/datasets/kddcup99.py:
```diff
@@ -11,32 +11,26 @@
-
-URL = ('http://archive.ics.uci.edu/ml/'
-       'machine-learning-databases/kddcup99-mld/kddcup.data.gz')
+URL10 = 'https://ndownloader.figshare.com/files/5976042'
```
I think it is now less clear what 10 means. Rename the variable to URL_10_PERCENT

jnothman commented on sklearn/datasets/rcv1.py:
```diff
@@ -124,16 +144,18 @@ def fetch_rcv1(data_home=None, subset='all', download_if_missing=True,
-        file_urls.append("%s_train.dat.gz" % URL)
         files = []
-        for file_url in file_urls:
+        for file_name, file_url in zip(FILE_NAMES, FILE_URLS):
```
If you already have zip, why not zip the checksums in too, rather than using a different data structure?

jnothman replied:
also, a needless inconsistency with lfw.

jnothman reviewed (commented):
I would appreciate if there were a bit more consistency across fetch functions as to how the checksum, URL and local path are related in code. We could of course just store this in a table globally, but I'm not sure that's so elegant either.

jnothman commented:
Good work, @nelson-liu

jnothman commented:
Would you be able to fix this up to merge for the upcoming release, @nelson-liu? Or should we get someone else to do the final touch-ups?

nelson-liu commented:
sorry for being so flaky --- it'd be good to get someone else to finish the last bits. I'll see if I can pitch in, though...

ogrisel commented:
Closing in favor of #9240.
